arch/sim/src/sim/up_alsa.c

```diff
@@ -0,0 +1,731 @@
+  .start         = sim_audio_start,
+#ifndef CONFIG_AUDIO_EXCLUDE_STOP
+  .stop          = sim_audio_stop,
+#endif
+#ifndef CONFIG_AUDIO_EXCLUDE_PAUSE_RESUME
+  .pause         = sim_audio_pause,
+  .resume        = sim_audio_resume,
+#endif
+  .enqueuebuffer = sim_audio_enqueuebuffer,
+  .ioctl         = sim_audio_ioctl,
+  .reserve       = sim_audio_reserve,
+  .release       = sim_audio_release,
+};
+
+static sq_queue_t g_sim_audio;
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static int sim_audio_config_format(struct sim_audio_s *priv, snd_pcm_t *pcm)
+{
+  snd_pcm_hw_params_t *hw_params;
+  snd_pcm_uframes_t pframes;
+  snd_pcm_format_t format;
+  uint32_t total_size;
+  int ret;
+
+  switch (priv->bps)
+    {
+      case 8:
+        format = SND_PCM_FORMAT_S8;
+        break;
+      case 24:
+        format = SND_PCM_FORMAT_S24;
+        break;
+      case 32:
+        format = SND_PCM_FORMAT_S32;
+        break;
+      default:
+        format = SND_PCM_FORMAT_S16;
+        break;
+    }
+
+  ret = snd_pcm_hw_params_malloc(&hw_params);
+  if (ret < 0)
+    {
+      return ret;
+    }
+
+  ret = snd_pcm_hw_params_any(pcm, hw_params);
+  if (ret < 0)
+    {
+      goto fail;
+    }
+
+  ret = snd_pcm_hw_params_set_access(pcm, hw_params,
+                                     SND_PCM_ACCESS_RW_INTERLEAVED);
+  if (ret < 0)
+    {
+      goto fail;
+    }
+
+  ret = snd_pcm_hw_params_set_format(pcm, hw_params, format);
+  if (ret < 0)
+    {
+      goto fail;
+    }
+
+  ret = snd_pcm_hw_params_set_rate(pcm, hw_params,
+                                   priv->sample_rate, 0);
+  if (ret < 0)
+    {
+      goto fail;
+    }
+
+  ret = snd_pcm_hw_params_set_channels(pcm, hw_params,
+                                       priv->channels);
+  if (ret < 0)
+    {
+      goto fail;
+    }
+
+  total_size = priv->nbuffers * priv->buffer_size;
+
+  pframes = priv->buffer_size / priv->frame_size;
+  ret = snd_pcm_hw_params_set_period_size_near(pcm, hw_params,
+                                               &pframes, NULL);
+  if (ret < 0)
+    {
+      goto fail;
+    }
+
+  priv->buffer_size = pframes * priv->frame_size;
+  priv->nbuffers    = total_size / priv->buffer_size;
+  ret = snd_pcm_hw_params_set_periods_near(pcm, hw_params,
+                                           &priv->nbuffers, NULL);
+  if (ret < 0)
+    {
+      goto fail;
+    }
+
+  ret = snd_pcm_hw_params(pcm, hw_params);
+
+fail:
+  snd_pcm_hw_params_free(hw_params);
+  return ret;
+}
+
+static int sim_audio_open(struct sim_audio_s *priv)
+{
+  snd_pcm_t *pcm;
+  int direction;
+  int ret;
+
+  if (priv->pcm)
+    {
+      return 0;
+    }
+
+  direction = priv->playback ? SND_PCM_STREAM_PLAYBACK
+                             : SND_PCM_STREAM_CAPTURE;
+
+  ret = snd_pcm_open(&pcm, "default", direction, 0);
```

Comment:
Ah I guess this is actually where my comment should be. Don't we want to specify this?

Reply:
Do you want to change the device name? Ok, it depend on how many pcm instances we want to create. only one playback/capture is created in the current design, "default" should be the best choice in this case. On the other hand, we sim_audio_initialize need accept name argument and open the different sound device.

Reply:
I guess I don't care if we limit to one device, but I want to be able to at least specify it via a config. "default" is certainly the correct default choice.

Reply:
In ALSA design, user can modify alsa.conf to map "default" to the different sound hardware. That's why we don't make the device name configurable initially. If you really need this feature, we can add it.

Reply:
Just seems trivial to add the two kconfig values here and then not require people to modify their system for testing.  This makes it a lot easier to use an alsa loopback device.

Reply:
If you want to still merge fine, it just does not align with other sim hardware where the user can specify how it connects to to the host hardware.

Reply:
How about we merge this patch first? and  @GUIDINGLI provide a new patch to support configuration in tomorrow.